# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1533831547 14400
#      Thu Aug 09 12:19:07 2018 -0400
# Branch stable
# Node ID b5dc88246c02467ec58900d30b3db3806b72d9fa
# Parent  092bb49b69624457e9b60c5f7b8711712e848311
disable compound binary operator optimization (bug #54465)

* oct-parse.in.yy (base_parser::make_binary_op):
Don't recognize compound binary operators here.

* parser.tst: New tests.

diff --git a/libinterp/parse-tree/oct-parse.in.yy b/libinterp/parse-tree/oct-parse.in.yy
--- a/libinterp/parse-tree/oct-parse.in.yy
+++ b/libinterp/parse-tree/oct-parse.in.yy
@@ -2641,17 +2641,17 @@ namespace octave
       default:
         panic_impossible ();
         break;
       }
 
     int l = tok_val->line ();
     int c = tok_val->column ();
 
-    return maybe_compound_binary_expression (op1, op2, l, c, t);
+    return new tree_binary_expression (op1, op2, l, c, t);
   }
 
   // Build a boolean expression.
 
   tree_expression *
   base_parser::make_boolean_op (int op, tree_expression *op1,
                                 token *tok_val, tree_expression *op2)
   {
diff --git a/test/parser.tst b/test/parser.tst
--- a/test/parser.tst
+++ b/test/parser.tst
@@ -332,8 +332,23 @@
 %! assert (f (1), 1);
 %! assert (f (1, 2), 2);
 
 ## FIXME: We need a sequence of concatenation tests since this seems
 ##        to be a frequently reported source of incompatibilities w/Matlab
 ## Check concatenation of empty char matrices (bug #52542)
 %!assert (double ([char(ones(0,3)); 'A']), 65)
 %!assert (double ([char(ones(0,3)); char(ones(2,0)); 'A']), 65)
+
+%!test <*54465>
+%! x = 0;
+%! if (! 0 || (x = 1))
+%! endif
+%! assert (x == 0);
+
+%!function x = matlab_short_circuit ()
+%!  x = 0;
+%!  if (! 0 | (x = 1))
+%!  endif
+%!endfunction
+%!test <*54465>
+%! fail ("x = matlab_short_circuit ()", "warning", "Matlab-style short-circuit operation performed for operator |");
+%! assert (x == 0);
